### PR TITLE
[WIP] Uniform polygon sampling

### PR DESCRIPTION
**Pull request by OwenGraves**
I'd like to implement uniform polygon sampling. I've added uniform triangle sampling, but am still working on the other methods.

Any and all feedback, criticism, or best practices information would be much appreciated! 😄

### CODE REVIEW

**Comment by Lemmih** [comment]
Nice!
You can split a polygon into triangles by using the triangulation module and this code snippet:

```haskell
import Algorithms.Geometry.PolygonTriangulation.Triangulate

polygonTriangles :: (Ord r, Fractional r) => Polygon t p r -> [SimplePolygon p r]
polygonTriangles p =
    map (view core) $
    map (`rawFacePolygon` g) $
    map fst $ V.toList (internalFaces g)
  where
    g = triangulate' Proxy p
```

File: hgeometry/src/Algorithms/Geometry/UniformPolygonSampling.hs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+-- | O(n log n)
+samplePolygon :: (RandomGen g, Random r, Fractional r) => Polygon t p r -> Rand g (Point 2 r)
+samplePolygon = error "not implemented yet"
+
+-- | O(1)
+sampleTriangle :: (RandomGen g, Random r, Fractional r, Ord r) => Triangle 2 p r -> Rand g (Point 2 r)
+sampleTriangle (Triangle v1 v2 v3) = do
+  a' <- getRandomR (0, 1)
+  b' <- getRandomR (0, 1)
+  let (a, b) = if a' + b' > 1 then (1 - a', 1 - b') else (a', b') in
```
Comment (Lemmih) [inline hgeometry/src/Algorithms/Geometry/UniformPolygonSampling.hs:21]: You can drop the `in` part of the let-expression and keep the `return` on the same level of indentation.

**Comment by OwenGraves** [comment]
Not sure if the animation is in the right place (or setting files modified correctly) and the animation could be beautified, but it seems to work!

**Comment by noinia** [comment]
Does this strategy of sampling triangles really produce a uniform sampling? Intuitively it seems similar to sampling in a disk by taking an uniform sample of polar coordinates. However that does *not* produce an uniform sampling in the disk, as you are more likely to pick points near the center this way. This type of sampling seems very similar to that, so I would like to see more evidence that this produces the right result.

A simple alternative strategy is to simply sample uniformly at random from the boundingbox of the triangle and resample if it turns out that point does not lie in the triangle (but that is a bit slower/may require more random bits).

**Comment by Lemmih** [comment]
The sampling /is/ uniform. We use the triangle to create a parallelogram which can be sampled uniformly. A sampled point has a 50% chance of being outside of the original triangle. If the point is outside, we reflect it over the triangle edge such that it ends up inside the triangle. Thus we have uniform sampling in O(1) steps.

This article describes the approach in detail: https://blogs.sas.com/content/iml/2020/10/19/random-points-in-triangle.html

**Comment by Lemmih** [comment]
I'll document the algorithm before merging this PR.

**Comment by Lemmih** [comment]
Hi @OwenGraves,
I've modified your animation to use text polygons:

https://user-images.githubusercontent.com/1443107/111307837-f0502c80-8694-11eb-8273-4cb220dbf11d.mp4

There are still a few TODO items left but I think we can handle those in another PR and merge this one.

**Comment by Lemmih** [comment]
TODO items:
 - [ ] `O(log n)` sampling using a fingertree.
 - [ ] Uniform sampling of multiple polygons at once.
 - [ ] Drop the `Real r` constraint (by not using `fromList`).

**Comment by Lemmih** [comment]
On the website: https://hgeometry.org/#uniformsampling

**Comment by OwenGraves** [comment]
Awesome, that animation looks really cool!

**Comment by OwenGraves** [comment]
Another TODO item could be to implement the [Alias method](https://en.wikipedia.org/wiki/Alias_method) since it looks like it might be a way to do this sampling in `O(1)` after the initial setup.

**Comment by noinia** [comment]
I've ported these results to the new hgeometry version in #238 , so I'm closing this one :).

**Pull request closed without merging**